File: frontend-mobile/lib/services/Gemified/problem_solving_lesson_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

import '../api_config.dart';

/// Problem Solving Lesson Service
/// Matches backend:
/// POST   /chromabloom/problem-solving-lessons
/// GET    /chromabloom/problem-solving-lessons
/// GET    /chromabloom/problem-solving-lessons/:id
/// PUT    /chromabloom/problem-solving-lessons/:id
/// DELETE /chromabloom/problem-solving-lessons/:id
class ProblemSolvingLessonService {
  ProblemSolvingLessonService._();

  // TODO: change to your server IP / domain
  static String get _baseUrl => ApiConfig.baseUrl; // Android emulator
  // static const String _baseUrl = "http://localhost:5000"; // Flutter web
  // static const String _baseUrl = "http://192.168.x.x:5000"; // real device WiFi

  static const String _path = "/chromabloom/problem-solving-lessons";

  static Map<String, String> _headers() => const {
        "Content-Type": "application/json",
        "Accept": "application/json",
      };

  // -----------------------------
  // Helpers
  // -----------------------------
  static Exception _errFromResponse(http.Response res) {
    try {
      final body = jsonDecode(res.body);
      final msg = body is Map && body["message"] != null
          ? body["message"].toString()
          : "Request failed (${res.statusCode})";
      return Exception(msg);
    } catch (_) {
      return Exception("Request failed (${res.statusCode})");
    }
  }

  /// Mini tutorial item builder
  /// tip_number: Number (required)
  /// tip_content: String (required)
  static Map<String, dynamic> miniTutorial({
    required int tipNumber,
    required String tipContent,
  }) {
    return {
      "tip_number": tipNumber,
      "tip_content": tipContent,
    };
  }

  // -----------------------------
  // CRUD
  // -----------------------------

  /// Create a lesson
  /// Backend expects:
  /// { title, description, difficulty_level, miniTutorialsName, miniTutorials }
  ///
  /// miniTutorials can be an array OR JSON string (your controller supports both).
  static Future<Map<String, dynamic>> createLesson({
    required String title,
    required String description,
    required String difficultyLevel, // Beginner | Intermediate | Advanced
    String? miniTutorialsName,
    List<Map<String, dynamic>>? miniTutorials,
    bool sendMiniTutorialsAsString = false, // optional
  }) async {
    final uri = Uri.parse("$_baseUrl$_path");

    final payload = <String, dynamic>{
      "title": title,
      "description": description,
      "difficulty_level": difficultyLevel,
      if (miniTutorialsName != null) "miniTutorialsName": miniTutorialsName,
      if (miniTutorials != null)
        "miniTutorials": sendMiniTutorialsAsString
            ? jsonEncode(miniTutorials)
            : miniTutorials,
    };

    final res = await http
        .post(uri, headers: _headers(), body: jsonEncode(payload))
        .timeout(const Duration(seconds: 15));

    if (res.statusCode == 201 || res.statusCode == 200) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    }
    throw _errFromResponse(res);
  }

  /// Get all lessons
  static Future<List<dynamic>> getAllLessons() async {
    final uri = Uri.parse("$_baseUrl$_path");

    final res = await http
        .get(uri, headers: _headers())
        .timeout(const Duration(seconds: 15));

    if (res.statusCode >= 200 && res.statusCode < 300) {
      final body = jsonDecode(res.body) as Map<String, dynamic>;
      return (body["data"] as List<dynamic>? ?? []);
    }
    throw _errFromResponse(res);
  }

  /// Get lesson by ID (LP-0001)
  static Future<Map<String, dynamic>> getLessonById(String id) async {
    final uri = Uri.parse("$_baseUrl$_path/$id");

    final res = await http
        .get(uri, headers: _headers())
        .timeout(const Duration(seconds: 15));

    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    }
    throw _errFromResponse(res);
  }

  /// Update lesson by ID
  /// Any field can be updated (send only what you want to change)
  static Future<Map<String, dynamic>> updateLesson({
    required String id,
    String? title,
    String? description,
    String? difficultyLevel, // Beginner | Intermediate | Advanced
    String? miniTutorialsName,
    List<Map<String, dynamic>>? miniTutorials,
    bool sendMiniTutorialsAsString = false,
  }) async {
    final uri = Uri.parse("$_baseUrl$_path/$id");

    final payload = <String, dynamic>{
      if (title != null) "title": title,
      if (description != null) "description": description,
      if (difficultyLevel != null) "difficulty_level": difficultyLevel,
      if (miniTutorialsName != null) "miniTutorialsName": miniTutorialsName,
      if (miniTutorials != null)
        "miniTutorials": sendMiniTutorialsAsString
            ? jsonEncode(miniTutorials)
            : miniTutorials,
    };

    final res = await http
        .put(uri, headers: _headers(), body: jsonEncode(payload))
        .timeout(const Duration(seconds: 15));

    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    }
    throw _errFromResponse(res);
  }

  /// Delete lesson by ID
  static Future<Map<String, dynamic>> deleteLesson(String id) async {
    final uri = Uri.parse("$_baseUrl$_path/$id");

    final res = await http
        .delete(uri, headers: _headers())
        .timeout(const Duration(seconds: 15));

    if (res.statusCode >= 200 && res.statusCode < 300) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    }
    throw _errFromResponse(res);
  }
}

File: frontend-web/src/services/Gemified_Knowledge_Builder/problemSolvingLessonService.js
// src/services/problemSolvingLessonService.js
import axios from "axios";

// Put your backend base URL in .env (recommended)
// VITE_API_BASE_URL=http://localhost:5000
const API_BASE =  "https://chromabloom-backend.onrender.com";

// Your router is mounted like:
// app.use("/chromabloom/problem-solving-lessons", router)
const ENDPOINT = `${API_BASE}/chromabloom/problem-solving-lessons`;

// ---------- helpers ----------
const normalizeMiniTutorials = (miniTutorials) => {
  // Allow:
  // - array (recommended)
  // - stringified JSON array
  // - undefined/null
  if (miniTutorials === undefined || miniTutorials === null) return undefined;

  // If already an array -> ok
  if (Array.isArray(miniTutorials)) return miniTutorials;

  // If string -> try parse
  if (typeof miniTutorials === "string") {
    try {
      const parsed = JSON.parse(miniTutorials);
      return parsed;
    } catch {
      // Keep as string (backend will throw a nice error)
      return miniTutorials;
    }
  }

  // otherwise send as-is (backend will validate)
  return miniTutorials;
};

// ---------- API calls ----------
export const ProblemSolvingLessonService = {
  // CREATE
  // payload: { title, description, difficulty_level, miniTutorialsName, miniTutorials: [] }
  create: async (payload) => {
    const body = {
      ...payload,
      miniTutorials: normalizeMiniTutorials(payload?.miniTutorials),
    };

    const res = await axios.post(ENDPOINT, body, {
      headers: { "Content-Type": "application/json" },
    });
    return res.data; // { success, data }
  },

  // GET ALL
  getAll: async () => {
    const res = await axios.get(ENDPOINT);
    return res.data; // { success, data: [] }
  },

  // GET BY ID (LP-0001)
  getById: async (id) => {
    const res = await axios.get(`${ENDPOINT}/${id}`);
    return res.data; // { success, data }
  },

  // UPDATE
  // payload can be partial
  update: async (id, payload) => {
    const body = {
      ...payload,
      miniTutorials: normalizeMiniTutorials(payload?.miniTutorials),
    };

    const res = await axios.put(`${ENDPOINT}/${id}`, body, {
      headers: { "Content-Type": "application/json" },
    });
    return res.data; // { success, data }
  },

  // DELETE
  remove: async (id) => {
    const res = await axios.delete(`${ENDPOINT}/${id}`);
    return res.data; // { success, message }
  },
};

export default ProblemSolvingLessonService;
